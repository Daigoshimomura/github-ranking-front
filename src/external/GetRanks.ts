import fetcher from "../util/Fetcher";
import { RankgByLanguageRes } from "./data/RanksByLanguageRes";

async function getRanks() {
  const endpoint = process.env.REACT_APP_BACKEND_ENDPOINT as string;
  const url = "/github-ranking/ranking";
  return await fetcher<RankgByLanguageRes>(endpoint + url, {
    method: "GET"
  });
}
const GetRanks = {
  getRanks: getRanks
};
export default GetRanks;
