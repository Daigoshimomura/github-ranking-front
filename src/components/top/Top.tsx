import React, { ChangeEvent } from "react";
import Button from "@material-ui/core/Button";
import { TextField, Container, Box, Grid, Snackbar } from "@material-ui/core";
import { Header } from "../parts/Header";
import { RouteComponentProps } from "react-router-dom";
import GetRanks from "../../external/GetRanks";
import { RankgByLanguageRes } from "../../external/data/RanksByLanguageRes";
import {
  JAVASCRIPT,
  GENERAL,
  PYTHON,
  JAVA,
  CSHARP,
  PHP,
  C_PLUS,
  RUBY,
  GO
} from "../../const/Language";
import { RankingList } from "../parts/RankingList";
import Alert from "@material-ui/lab/Alert";
import topImage from "../../static/top.png";

interface TopProps extends RouteComponentProps<{}> {
  screenName: string;
}

export interface TopState {
  userId: string;
  dialogOpen: boolean;
  ranking: RankgByLanguageRes;
}

export class Top extends React.Component<TopProps, TopState> {
  constructor(props: TopProps, state: TopState) {
    super(props);
    this.state = {
      userId: "",
      dialogOpen: false,
      ranking: {
        rankByLanguages: [
          { language: GENERAL, userInfomations: [] },
          { language: JAVASCRIPT, userInfomations: [] },
          { language: PYTHON, userInfomations: [] },
          { language: JAVA, userInfomations: [] },
          { language: CSHARP, userInfomations: [] },
          { language: PHP, userInfomations: [] },
          { language: C_PLUS, userInfomations: [] },
          { language: RUBY, userInfomations: [] },
          { language: GO, userInfomations: [] }
        ]
      }
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.moveDetailLogic = this.moveDetailLogic.bind(this);
  }

  dialogClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    this.setState({ dialogOpen: false });
  };

  componentDidMount() {
    this.getRanking();
  }

  getRanking() {
    GetRanks.getRanks().then(res => {
      this.setState({ userId: "", ranking: res });
    });
  }

  handleClick() {
    if (this.state.userId === "") {
      this.setState({ dialogOpen: true });
      return;
    }
    this.moveDetailLogic(this.state.userId);
  }

  moveDetailLogic(userId: String) {
    this.props.history.push({
      pathname: "/detail",
      state: { userId: userId }
    });
  }

  handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      userId: e.target.value
    });
  }

  render() {
    const general = this.state.ranking.rankByLanguages.filter(
      t => t.language === GENERAL
    )[0];
    const javascript = this.state.ranking.rankByLanguages.filter(
      t => t.language === JAVASCRIPT
    )[0];
    const python = this.state.ranking.rankByLanguages.filter(
      t => t.language === PYTHON
    )[0];
    const java = this.state.ranking.rankByLanguages.filter(
      t => t.language === JAVA
    )[0];
    const csharp = this.state.ranking.rankByLanguages.filter(
      t => t.language === CSHARP
    )[0];
    const php = this.state.ranking.rankByLanguages.filter(
      t => t.language === PHP
    )[0];
    const cplus = this.state.ranking.rankByLanguages.filter(
      t => t.language === C_PLUS
    )[0];
    const ruby = this.state.ranking.rankByLanguages.filter(
      t => t.language === RUBY
    )[0];
    const go = this.state.ranking.rankByLanguages.filter(
      t => t.language === GO
    )[0];

    return (
      <div>
        <Container>
          <Header />
          <Box textAlign="center" m={3}>
            <img src={topImage} alt={""} width={400} height={120}></img>
          </Box>
          <Box textAlign="center">
            <TextField
              id="standard-basic"
              label="GitHub ID"
              value={this.state.userId}
              onChange={this.handleChange}
            />
          </Box>
          <Box textAlign="center" mt={2}>
            <Button
              variant="contained"
              color="primary"
              onClick={this.handleClick}
            >
              SHOW PROFILE
            </Button>
          </Box>
          <Box m={10}>
            <Grid container spacing={2}>
              <RankingList
                ranksByLanguage={general}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={javascript}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={python}
                itemClickCallback={this.moveDetailLogic}
              />
            </Grid>
          </Box>
          <Box m={10}>
            <Grid container spacing={2}>
              <RankingList
                ranksByLanguage={java}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={csharp}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={php}
                itemClickCallback={this.moveDetailLogic}
              />
            </Grid>
          </Box>
          <Box m={10}>
            <Grid container spacing={2}>
              <RankingList
                ranksByLanguage={cplus}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={ruby}
                itemClickCallback={this.moveDetailLogic}
              />
              <RankingList
                ranksByLanguage={go}
                itemClickCallback={this.moveDetailLogic}
              />
            </Grid>
          </Box>
          <Snackbar open={this.state.dialogOpen} autoHideDuration={6000}>
            <Alert onClose={this.dialogClose} severity="error">
              please fill in the value !
            </Alert>
          </Snackbar>
        </Container>
      </div>
    );
  }
}
