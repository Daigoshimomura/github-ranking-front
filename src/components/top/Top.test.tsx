import { mount } from "enzyme";
import { Top, TopState } from "./Top";
import { routerTestProps } from "../../util/test/RoutComponentMock";
import { createLocation, createMemoryHistory } from "history";
import React from "react";
import { Snackbar } from "@material-ui/core";
import GetRanks from "../../external/GetRanks";
import renderer from "react-test-renderer";

function setUp(state: TopState) {
  const { match } = routerTestProps("/top/", {});
  const location = createLocation<TopState>(match.url);
  location.state = state;
  const history = createMemoryHistory<TopState>();

  const wrapper = mount<Top>(
    <Top
      match={match}
      location={location}
      history={history}
      screenName={"top"}
    />
  );

  return wrapper;
}

describe("Top", () => {
  it("snapshot", () => {
    const { match } = routerTestProps("/top/", {});
    const location = createLocation<TopState>(match.url);
    location.state = {
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    };
    const history = createMemoryHistory<TopState>();
    const tree = renderer
      .create(
        <Top
          match={match}
          location={location}
          history={history}
          screenName={"top"}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("レンダリングの確認", () => {
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    expect(wrapper.find(Snackbar).length).toBe(1);
  });

  test("componentDidMountの確認", () => {
    const getRanking = jest.spyOn(Top.prototype, "getRanking");
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    expect(getRanking).toHaveBeenCalled();
  });

  test("dialogCloseの確認", () => {
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    const comp: Top = wrapper.instance();
    comp.dialogClose();
    expect(comp.state.dialogOpen).toBe(false);
  });

  test("getRankingの確認", () => {
    const getRanksMock = jest
      .fn()
      .mockImplementation(() => Promise.resolve({}));
    GetRanks.getRanks = getRanksMock;
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    const comp: Top = wrapper.instance();
    comp.getRanking();
    expect(getRanksMock).toHaveBeenCalled();
  });

  test("handleClickの確認", () => {
    const moveDetailLogic = jest.spyOn(Top.prototype, "moveDetailLogic");
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    const comp: Top = wrapper.instance();
    comp.setState({ userId: "userId" });
    comp.handleClick();
    expect(moveDetailLogic).toHaveBeenCalled();
  });

  test("moveDetailLogicの確認", () => {
    const mock = jest.fn();
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    const comp: Top = wrapper.instance();
    comp.props.history.push = mock;
    comp.moveDetailLogic("userId");
    expect(mock).toHaveBeenCalled();
  });

  test("handleChangeの確認", () => {
    const wrapper = setUp({
      userId: "userId",
      dialogOpen: true,
      ranking: {
        rankByLanguages: []
      }
    });
    const comp: Top = wrapper.instance();
    wrapper.find("input").simulate("change", { target: { value: "hogehoge" } });

    expect(comp.state.userId).toBe("hogehoge");
  });
});
