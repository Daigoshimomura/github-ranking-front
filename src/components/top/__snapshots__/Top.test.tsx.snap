// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top snapshot 1`] = `
<div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
        <h6
          className="MuiTypography-root MuiTypography-h6"
        >
          GitHub Ranking
        </h6>
      </div>
    </header>
    <div
      className="MuiBox-root MuiBox-root-116"
    >
      <img
        alt=""
        height={120}
        src="top.png"
        width={400}
      />
    </div>
    <div
      className="MuiBox-root MuiBox-root-117"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="standard-basic"
          id="standard-basic-label"
        >
          GitHub ID
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="standard-basic"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root MuiBox-root-175"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          SHOW PROFILE
        </span>
      </button>
    </div>
    <div
      className="MuiBox-root MuiBox-root-205"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            General
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            JavaScript
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            Python
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root MuiBox-root-313"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            Java
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            C#
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            PHP
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root MuiBox-root-314"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            C++
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            Ruby
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            Go
          </h6>
          <div>
            <ul
              className="MuiList-root MuiList-padding"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
